docs/source/miscellaneous/doc/adr/0014-cassandra-targeted-sweep.md

```diff
@@ -0,0 +1,321 @@
+3. Check the *commit timestamp* of the transaction.
+  - If the transaction is not committed, abort it.
+  - If the transaction is aborted, delete the write from the KVS and pop the queue and read the next entry.
+  - If the transaction is committed at a timestamp greater or equal to the *sweep timestamp*, pause and try later.
+4. Otherwise, insert a ranged tombstone as follows:
+  - If the strategy is conservative: write a garbage deletion sentinel, then put a ranged tombstone deleting all versions of the
+  cell between 0 and write timestamp - 1, not deleting the sentinel or the write.
+  - If the strategy is thorough
+    - If the write was a tombstone, then put a ranged tombstone deleting all versions of that cell between -1 and write timestamp,
+    deleting both a potentially existing sentinel and the write.
+    - Otherwise, put a ranged tombstone deleting all versions of that cell between -1 and write timestamp - 1, deleting a
+    potentially existing sentinel, but not the write.
+5. Pop the queue and read the next entry.
+
+For a detailed implementation of targeted sweep and the targeted sweep queue, refer to the implementation section below.
+
+## Consequences
+
+### Benefits
+- Targeted sweep can perform sweeps several orders of magnitude times faster than legacy sweep (this has only been verified on Cassandra so far).
+- The load on the underlying KVS is significantly reduced (this has only been verified on Cassandra so far).
+- The order of sweeping is more fair and does not suffer from issues caused by frequently appending new rows to the end of a table.
+
+### Drawbacks
+- Added overhead to committing transactions, as the information must be persisted to the sweep queue as part of the commit. Note
+  that this has not caused significant regression in our benchmarks.
+
+## Implementation
+
+Since we assume that threads may die at any moment, for most of the implementation details below, correct ordering is crucial.
+Before we describe the targeted sweep schema, we will define the necessary terms. A **Fine Timestamp Partition** of a timestamp *ts*
+is *ts* divided by *50_000*, using integer division. A fine timestamp partition of a write is the fine
+timestamp partition of the start timestamp of the transaction that wrote it. A **Coarse Timestamp Partition** is analogous to the
+fine partition, except the number divided by is *10_000_000*. A **Sweep Strategy** is a table-level property specifying how the
+sweep timestamp should be calculated, if we need deletion sentinels, and whether the latest write should be swept away as well in
+case it is a delete. We also use a sharding strategy
+to enable better parallelisation of targeted sweep. The maximum number of **Shards** that is supported is *256*, effectively
+splitting the queue into that number of disjoint queues. Note that once that the number of shards is increased, it cannot be
+lowered again.
+
+### SWEEPABLECELLS
```

Comment:
I don't really like the all caps here - it's obvious where the word boundaries are in this example, but in general it makes parsing tricky (e.g. internal transformation product's tables)

Reply:
Is this better? Looks a bit weird, but seems better than the other options

Reply:
Yep! That said, I'm not sure what was wrong with `SweepableCells` (was this avoided for confusion with the java class maybe)?